metadata_harvester_cli.py

```diff
@@ -1,30 +1,83 @@
+def get_last_harvest_date():
+    """Only successful harvests are logged. This function gets the last harvesting date from the log.
+    :param filename: string value of a file name
+    :return: date in last line
+    """
+    log_file_path = "harvester.log"
+    try:
+        with open(log_file_path, "r") as file:
+            lines = file.readlines()
+            if lines:
+                log_date = lines[-1].split()[0]
+                return log_date
+    except FileNotFoundError:
+        return None
+    
+def retrieve_metadata_content(url="https://kielipankki.fi/md_api/que"):
     """
-    Fetch metadata records and transform them to JSON.
+    Fetches metadata records since the last logged harvest. If date is missing, all records are fetched.
+    :param url: string value of a url
+    :return: dictionary of mapped data
     """
-    api = PMH_API(url)
-    metadata_contents = api.get_all_metadata_records()
-    for metadata_content in metadata_contents:
-        lxml_record = etree.fromstring(etree.tostring(metadata_content.xml))
-        metadata_record = MSRecordParser(lxml_record)
-        click.echo(metadata_record.json_converter())
+    try:
+        api = PMH_API(url)
+        all_mapped_data_dict = {}
+        if get_last_harvest_date():
+            metadata_contents = api.get_changed_records_from_last_harvest(get_last_harvest_date())
+        else:
+            metadata_contents = api.get_all_metadata_records()
+
+        for metadata_content in metadata_contents:
+            lxml_record = etree.fromstring(etree.tostring(metadata_content.xml))
+            metadata_record = MSRecordParser(lxml_record)
+            if metadata_record.check_pid_exists():
+                if metadata_record.check_resourcetype_corpus():
+                    pid = metadata_record.get_identifier("//info:identificationInfo/info:identifier/text()")
+                    all_mapped_data_dict[pid] = metadata_record.to_dict()
+        return all_mapped_data_dict
+    except:
+        raise
+
+def send_data_to_metax(all_mapped_data_dict):
+    """
+    Make PUT and POST requests based on changes and existance of PIDs in Metax.
+    :param all_mapped_data_dict: a dictionary of mapped data
+    """
+    if all_mapped_data_dict:
+        try:
+            for pid in all_mapped_data_dict.keys():
+                dataset_dict = all_mapped_data_dict[pid]
+                if metax_api.check_if_dataset_record_in_datacatalog(pid):
+                    metax_dataset_id = metax_api.get_dataset_record_metax_id(pid)
+                    metax_api.update_dataset(metax_dataset_id, dataset_dict)
+                else:
+                    metax_api.create_dataset(dataset_dict)
+        except:
+            raise
+    else:
+        pass
+
 
 if __name__ == "__main__":
-    retrieve_metadata_content()
\ No newline at end of file
+    try:
+        send_data_to_metax(retrieve_metadata_content())
+        logger_harvester.info("Success")
```

Comment:
Implemented this now so that it logs the start time of a harvest. The code goes through the log until it finds a successful harvest and then uses that harvest's start time for the next round of harvesting.

Reply:
Nice, looks good!

Now we parse the log separately for logging purpose and for the actual record fetching. This is a bit risky, as it's very easy to change one of the hard-coded file names without remembering to change the other one, resulting in logs saying one thing ("updated since forever") and the actual harvesting having done something else ("updated since last week") which can be hard to debug. Now that we already have the datetime at hand, I suggest we pass it directly to `records_to_dict`.